treesit-fold.el

```diff
@@ -558,6 +582,81 @@ If the current syntax node is not foldable, do nothing."
           t)
       (treesit-fold-close))))
 
+(defun treesit-fold--node-cycle-state (&optional node)
+  "Return the cycle state of current heading.
+Return either hide-all, headings-only, or show-all."
+  (let (nodes)
+    (let* ((treesit-fold-indicators-mode)
+           (treesit-fold-on-fold-hook)
+           (node (or node (treesit-fold--foldable-node-at-pos)))
+           (children (treesit-node-children node))
+           (child (nth 0 children))
+           (beg (treesit-fold--get-fold-range node))
+           (nodes (treesit-fold--overlays-in
+                   'invisible
+                   'treesit-fold (treesit-node-start child) (treesit-node-end node)))
+           (first (cl-find-if (lambda (ov)
+                                (equal (overlay-start ov) (car beg)))
+                              nodes)))
+      (cond ((null nodes) 'show-all)
+            (first  'hide-all)
+            (t 'headings-only)))))
+
+(defun treesit-fold-node-cycle (&optional node)
+  "Cycle visibility state of the current node.
+
+This cycles the visibility of the current node and children between
+`hide all', `headings only' and `show all'.
+
+`Hide all' means hide everything in parent node.
+`Headings only' means show child nodes, but not their bodies.
+`Show all' means show all the child nodes and their bodies.
+
+If non-nil, EVENT should be a mouse event."
+  (interactive)
+  (let* ((node (or node (treesit-fold--foldable-node-at-pos))))
+    (when node
+      (pcase (treesit-fold--node-cycle-state node)
+        ('hide-all
+         (treesit-fold-open-recursively t node)
+         (if (treesit-fold--node-nodes node t)
+             (message "Only headings")
+           (message "Show all")))
+        ('headings-only
+         (treesit-fold-open-recursively nil node)
+         (message "Show all"))
+        ('show-all
+         (treesit-fold-close-recursively nil node)
+         (message "Hide all"))))))
+
+(defun treesit-fold-buffer-cycle ()
+  "Cycle visibility state of the buffer.
+
+This cycles the visibility of the all nodes and children between
+`hide all', `headings only' and `show all'.
+
+`Hide all' means hide everything in parent node.
+`Headings only' means show child nodes, but not their bodies.
+`Show all' means show all the child nodes and their bodies.
+
+If non-nil, EVENT should be a mouse event."
+  (interactive)
+  (let* ((children (treesit-node-children (treesit-buffer-root-node)))
+         (state (seq-map #'treesit-fold--node-cycle-state children)))
+    (pcase state
+      ((pred (member 'hide-all))
+       (seq-map (lambda (n)
+                  (treesit-fold-open-recursively t n))
+                children))
+      ((pred (member 'headings-only))
+       (seq-map (lambda (n)
+                  (treesit-fold-open-recursively nil n))
+                children))
+      ((pred (member 'show-all))
+       (seq-map (lambda (n)
+                  (treesit-fold-close-recursively nil n))
+                children)))))
```

Comment:
You might want to autoload these commands. 

Can you update the README according to the newly added commands?